infinispan-embedded/src/main/java/org/hibernate/ogm/datastore/infinispan/persistencestrategy/counter/ClusteredCounterCommand.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.datastore.infinispan.persistencestrategy.counter;
+
+import java.util.concurrent.ExecutionException;
+
+import org.hibernate.HibernateException;
+import org.hibernate.ogm.dialect.spi.NextValueRequest;
+import org.hibernate.ogm.model.key.spi.IdSourceKeyMetadata;
+
+import org.infinispan.counter.EmbeddedCounterManagerFactory;
+import org.infinispan.counter.api.CounterConfiguration;
+import org.infinispan.counter.api.CounterManager;
+import org.infinispan.counter.api.CounterType;
+import org.infinispan.counter.api.Storage;
+import org.infinispan.counter.api.StrongCounter;
+import org.infinispan.manager.EmbeddedCacheManager;
+
+/**
+ * @author Fabio Massimo Ercoli (C) 2017 Red Hat Inc.
+ */
+public class ClusteredCounterCommand {
+
+	public Number nextValue(EmbeddedCacheManager cacheManager, NextValueRequest request) {
+
+		CounterManager counterManager = EmbeddedCounterManagerFactory.asCounterManager( cacheManager );
```

Comment:
Could you extract all the "preparation" code so that it's run only once?

Then we can reuse the `ClusteredCounterCommand` for a specific counter. Make sure to have one of these created upfront for each sequence source we need.